external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,517 @@
+    }
+
+    private ConsumerRecords<K, V> pollKafkaBroker() {
+        final ConsumerRecords<K, V> consumerRecords = kafkaConsumer.poll(kafkaSpoutConfig.getPollTimeoutMs());
+        final int numPolledRecords = consumerRecords.count();
+        numUncommittedOffsets+= numPolledRecords;
+        LOG.debug("Polled [{}] records from Kafka. NumUncommittedOffsets=[{}]", numPolledRecords, numUncommittedOffsets);
+        return consumerRecords;
+    }
+
+    private void emitTuples(ConsumerRecords<K, V> consumerRecords) {
+        for (TopicPartition tp : consumerRecords.partitions()) {
+            final Iterable<ConsumerRecord<K, V>> records = consumerRecords.records(tp.topic());
+
+            for (final ConsumerRecord<K, V> record : records) {
+                final List<Object> tuple = tupleBuilder.buildTuple(record, kafkaSpoutStreams);
+                final KafkaSpoutMessageId messageId = new KafkaSpoutMessageId(record, tuple);
+
+                kafkaSpoutStreams.emit(collector, messageId);           // emits one tuple per record
+                LOG.trace("Emitted tuple [{}] for record [{}]", tuple, record);
+            }
+        }
+    }
+
+    private void commitOffsetsForAckedTuples() {
+        // Find offsets that are ready to be committed for every topic partition
+        final Map<TopicPartition, OffsetAndMetadata> nextCommitOffsets = new HashMap<>();
+        for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+            final OffsetAndMetadata nextCommitOffset = tpOffset.getValue().findNextCommitOffset();
+            if (nextCommitOffset != null) {
+                nextCommitOffsets.put(tpOffset.getKey(), nextCommitOffset);
+            }
+        }
+
+        // Commit offsets that are ready to be committed for every topic partition
+        if (!nextCommitOffsets.isEmpty()) {
+            kafkaConsumer.commitSync(nextCommitOffsets);
+            LOG.debug("Offsets successfully committed to Kafka [{}]", nextCommitOffsets);
+            // Instead of iterating again, it would be possible to commit and update the state for each TopicPartition
+            // in the prior loop, but the multiple network calls should be more expensive than iterating twice over a small loop
+            for (Map.Entry<TopicPartition, OffsetEntry> tpOffset : acked.entrySet()) {
+                final OffsetEntry offsetEntry = tpOffset.getValue();
+                offsetEntry.commit(nextCommitOffsets.get(tpOffset.getKey()));
+            }
+        } else {
+            LOG.trace("No offsets to commit. {}", this);
+        }
+    }
+
+    // ======== Ack =======
+
+    @Override
+    public void ack(Object messageId) {
+        if (!consumerAutoCommitMode) {  // Only need to keep track of acked tuples if commits are not done automatically
+            final KafkaSpoutMessageId msgId = (KafkaSpoutMessageId) messageId;
+            acked.get(msgId.getTopicPartition()).add(msgId);
```

Comment:
So in non-auto commit mode, we keep adding message ids to the acked messages. That list is cleared only when the commit timer expires. Correct me if I am wrong but there is no cap on the number of messages in the acked?